DOC Document changes to CLI interaction (#571)
silverstripe/developer-docs: en/08_Changelogs/6.0.0.md

There is also a new [`PolyCommand`](api:SilverStripe\PolyExecution\PolyCommand) class which provides a standardised API for code that needs to be accessible from both an HTTP request and CLI. This is used for `BuildTask` and other code accessed via `/dev/*` as mentioned below.

See [`PolyCommand`](/developer_guides/cli/polycommand) for more details about the `PolyCommand` API.

#### Changes to `BuildTask` {#cli-buildtask}

##### Change to API {#cli-buildtask-api}

The [`BuildTask`](api:SilverStripe\Dev\BuildTask) class is now a subclass of `PolyCommand`.

As a result of this, any `BuildTask` implementations in your project or module will need to be updated. The upgrade will likely look like this in most cases:

```diff
 namespace App\Tasks;

-use SilverStripe\Control\Director;
-use SilverStripe\Control\HTTPRequest;
 use SilverStripe\Dev\BuildTask;
+use SilverStripe\PolyExecution\PolyOutput;
+use Symfony\Component\Console\Command\Command;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Input\InputOption;

 class MyCustomTask extends BuildTask;
 {
-    private static $segment = 'my-custom-task';

> +    protected static string $commandName = 'my-custom-task';

Reviewer: ```suggestion
+    protected static string $commandName = 'MyCustomTask';
```

Reviewer 2: Discussion about this is ongoing in the issue.

+        ];
+    }
 }
```

Note that you should no longer output "done" or some equivalent message at the end of the task. Any time a task finishes executing, the output will include a message about whether the task completed successfully or failed (based on the return value) and how long it took.

See [`PolyCommand`](/developer_guides/cli/polycommand#buildtask) for more details about the `BuildTask` API.

##### Change to names {#cli-buildtask-names}

Some tasks were relying on the FQCN instead of having an explicit name. This means the name used for both the URL and the CLI command were excessively long.

The way these are displayed in the new `sake tasks` list doesn't suit long names, so we have given explicit names to these tasks in order to make them shorter. If you have scripts or cron jobs that reference these you will need to update them to use the new name.

For example, you used to navigate to `/dev/tasks/<OLD_NAME>` or use `sake dev/tasks/<OLD_NAME>`. Now you will need to navigate to `/dev/tasks/<NEW_NAME>` or use `sake tasks:<NEW_NAME>`.

|class|old name|new name|
|---|---|---|

> |[`ContentReviewEmails`](api:SilverStripe\ContentReview\Tasks\ContentReviewEmails)|`SilverStripe-ContentReview-Tasks-ContentReviewEmails`|`content-review-emails`|
> |[`DeleteAllJobsTask`](api:Symbiote\QueuedJobs\Tasks\DeleteAllJobsTask)|`Symbiote-QueuedJobs-Tasks-DeleteAllJobsTask`|`delete-queued-jobs`|
> |[`MigrateContentToElement`](api:DNADesign\Elemental\Tasks\MigrateContentToElement)|`DNADesign-Elemental-Tasks-MigrateContentToElement`|`elemental-migrate-content`|
> |[`UserFormsColumnCleanTask`](api:SilverStripe\UserForms\Task\UserFormsColumnCleanTask)|`SilverStripe-UserForms-Task-UserFormsColumnCleanTask`|`userforms-column-clean`|
> |[`StaticCacheFullBuildTask`](api:SilverStripe\StaticPublishQueue\Task\StaticCacheFullBuildTask)|`SilverStripe-StaticPublishQueue-Task-StaticCacheFullBuildTask`|`static-cache-full-build`|

Reviewer: ```suggestion
|[`ContentReviewEmails`](api:SilverStripe\ContentReview\Tasks\ContentReviewEmails)|`SilverStripe-ContentReview-Tasks-ContentReviewEmails`|`ContentReviewEmails`|
|[`DeleteAllJobsTask`](api:Symbiote\QueuedJobs\Tasks\DeleteAllJobsTask)|`Symbiote-QueuedJobs-Tasks-DeleteAllJobsTask`|`DeleteAllJobsTask`|
|[`MigrateContentToElement`](api:DNADesign\Elemental\Tasks\MigrateContentToElement)|`DNADesign-Elemental-Tasks-MigrateContentToElement`|`MigrateContentToElement`|
|[`UserFormsColumnCleanTask`](api:SilverStripe\UserForms\Task\UserFormsColumnCleanTask)|`SilverStripe-UserForms-Task-UserFormsColumnCleanTask`|`UserFormsColumnCleanTask`|
|[`StaticCacheFullBuildTask`](api:SilverStripe\StaticPublishQueue\Task\StaticCacheFullBuildTask)|`SilverStripe-StaticPublishQueue-Task-StaticCacheFullBuildTask`|`StaticCacheFullBuildTask`|
```

Reviewer 2: Discussion about this is ongoing in the issue.

#### Changes to `/dev/*` actions {#cli-dev}

With the changes to `sake` come changes to the way `dev/*` actions are handled. Most of these are now subclasses of the new [`DevCommand`](api:SilverStripe\Dev\Command\DevCommand) class which is itself a subclass of `PolyCommand`.

If you have custom actions registered under `DevelopmentAdmin.registered_controllers` you'll need to update the YAML configuration for these. If you want them to be accessible via CLI, you'll also have to update the PHP code.

With the below example, there are two custom actions displayed in the list at `/dev`:

> - `/dev/my-action`: intended for use in the browser only, but you'd have to add custom logic in `init()` to disallow its use in CLI until now
> - `/dev/my-other-action`: intended for use both in CLI and in the browser.

Reviewer: ```suggestion
- `/dev/my-http-only-action`: intended for use in the browser only, but you'd have to add custom logic in `init()` to disallow its use in CLI until now
- `/dev/my-http-and-cli-action`: intended for use both in CLI and in the browser.
```

> - `/dev/my-action`: intended for use in the browser only, but you'd have to add custom logic in `init()` to disallow its use in CLI until now
> - `/dev/my-other-action`: intended for use both in CLI and in the browser.

Reviewer: ```suggestion
- `/dev/my-http-only-action`: intended for use in the browser only, but you'd have to add custom logic in `init()` to disallow its use in CLI until now
- `/dev/my-http-and-cli-action`: intended for use both in CLI and in the browser.
```

Reviewer 2: Done

For actions that should only be accessible in the browser, you only need to change how these are registered. Move them from `DevelopmentAdmin.registered_controllers` to the new [`DevelopmentAdmin.controllers`](api:SilverStripe\Dev\DevelopmentAdmin->controllers) configuration property.

> Controllers added to `DevelopmentAdmin.registered_controllers` can only be accessed via HTTP requests, so you can remove any logic around CLI usage.

Reviewer: ```suggestion
Controllers added to `DevelopmentAdmin.controllers` can only be accessed via HTTP requests, so you can remove any logic around CLI usage.
```

Reviewer 2: Done

> Controllers added to `DevelopmentAdmin.registered_controllers` can only be accessed via HTTP requests, so you can remove any logic around CLI usage.

Reviewer: ```suggestion
Controllers added to `DevelopmentAdmin.controllers` can only be accessed via HTTP requests, so you can remove any logic around CLI usage.
```

Reviewer 2: Duplicate of https://github.com/silverstripe/developer-docs/pull/571#discussion_r1770699036?

For actions that should be accessible in the browser *and* via CLI, you will need to change these from being a `Controller` to subclassing `DevCommand`. These get registered to the new [`DevelopmentAdmin.commands`](api:SilverStripe\Dev\DevelopmentAdmin->commands) configuration property

```diff
 SilverStripe\Dev\DevelopmentAdmin:
-  registered_controllers:
-    my-action:
-      controller: 'App\Dev\MyActionController'
-      links:
-        my-action: 'Perform my custom action in dev/my-action (do not run in CLI)'
-    my-other-action:
-      controller: 'App\Dev\MyOtherActionController'
-      links:
-        my-other-action: 'Perform my custom action in dev/my-other-action'
+  controllers:

> +    my-action:
> +      class: 'App\Dev\MyActionController'
> +      description: 'Perform my custom action in dev/my-action'
> +  commands:
> +    my-other-action: 'App\Dev\MyOtherActionCommand'

Reviewer: ```suggestion
+    my-http-only-action:
+      class: 'App\Dev\MyHttpOnlyActionController'
+      description: 'Perform my custom action in dev/my-http-only-action'
+  commands:
+    my-http-and-cli-action: 'App\Dev\MyHttpAndCliActionCommand'
```

Reviewer 2: Done

```

```diff
 namespace App\Dev;

-use SilverStripe\Control\Controller;
-use SilverStripe\Control\Director;
-use SilverStripe\Control\HTTPRequest;
+use SilverStripe\Dev\Command\DevCommand;
-use SilverStripe\Dev\DevelopmentAdmin;
+use SilverStripe\PolyExecution\PolyOutput;
-use SilverStripe\Security\Permission;
 use SilverStripe\Security\PermissionProvider;
-use SilverStripe\Security\Security;
+use Symfony\Component\Console\Command\Command;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Input\InputOption;

-class MyOtherActionController extends Controller implements PermissionProvider

> +class MyOtherActionController extends DevCommand implements PermissionProvider

Reviewer: ```suggestion
+class MyHttpAndCliActionController extends DevCommand implements PermissionProvider
```

Reviewer 2: Done - also updated `MyActionController` to `MyHttpOnlyActionController`

 {

> +    protected static string $commandName = 'app:my-other-action';
> +
> +    protected static string $description = 'Perform my custom action in dev/my-other-action or via sake app:my-other-action';

Reviewer: ```suggestion
+    protected static string $commandName = 'app:my-http-and-cli-action';
+
+    protected static string $description = 'Perform my custom action in dev/my-http-and-cli-action or via sake app:my-http-and-cli-action';
```

Reviewer 2: Done

+    {
+        return 'My other action';
+    }
+
-    protected function init(): void
-    {
-        parent::init();
-
-        if (!$this->canInit()) {
-            Security::permissionFailure($this);
-        }
-    }
-
-    public function index(HTTPRequest $request)
+    protected function execute(InputInterface $input, PolyOutput $output): int
     {
-        $someVar = $request->getVar('some-var');
+        $input->getOption('some-var');
-        if (Director::is_cli()) {
-            $body = "some output\n";
-        } else {
-            $body = "some output<br>\n";
-        }
+        $output->writeln('some output');
-
-        return $this->getResponse()->setBody($body);
+        return Command::SUCCESS;
     }

+    public function getOptions(): array
+    {
+        return [
+            new InputOption('some-var', null, InputOption::VALUE_NONE, 'some get variable'),
+        ];
+    }
-    public function canInit(): bool
-    {
-        return (
-            Director::isDev()
-            || (Director::is_cli() && DevelopmentAdmin::config()->get('allow_all_cli'))
-            || Permission::check('MY_CUSTOM_PERMISSION')
-        );
-    }

     // ...
 }
```

> You would now access the `/dev/my-action` action via an HTTP request only. The `/dev/my-other-action` action can be access via an HTTP request, or by using `sake app:my-other-action` on the command line.

Reviewer: ```suggestion
You would now access the `/dev/my-http-only-action` action via an HTTP request only. The `/dev/my-http-and-cli-action` action can be access via an HTTP request, or by using `sake app:my-http-and-cli-action` on the command line.
```

Reviewer 2: Done.... I think this is a duplicate suggestion? Or else you suggested other related changes way earlier and now it's hard to tell if this is a duplicate or just related.
Done in any case.

> You would now access the `/dev/my-action` action via an HTTP request only. The `/dev/my-other-action` action can be access via an HTTP request, or by using `sake app:my-other-action` on the command line.
> 
> The `some-var` option can be used in a query string when running the action via an HTTP request, or as a flag (e.g. `sake app:my-other-action --some-var`) in CLI.

Reviewer: ```suggestion
You would now access the `/dev/my-http-only-action` action via an HTTP request only. The `/dev/my-http-and-cli-action` action can be access via an HTTP request, or by using `sake app:my-http-and-cli-action` on the command line.

The `some-var` option can be used in a query string when running the action via an HTTP request, or as a flag (e.g. `sake app:my-http-and-cli-action --some-var`) in CLI.
```

Reviewer 2: Done

See [`PolyCommand`](/developer_guides/cli/polycommand#dev-commands) for more details about the `DevCommand` API.

#### `sake -start` and `sake -stop` have been removed {#cli-daemon}

Sake used to have functionality to make daemon processes for your application. This functionality was managed with `sake -start my-process` and `sake -stop my-process`.

We've removed this functionality. Please use an appropriate daemon tool such as `systemctl` to manage these instead.

> 

Reviewer: Need to mention `dont_populate` being renamed to `no-populate`

Reviewer 2: That's not related to `sake -start` and `sake -stop`, but I've added a note about it to the relevant place.
